src/profiler.ts

```diff
@@ -13,57 +13,236 @@
       tag: pjson.name
     });
+
+    // TODO: enable heap profiling once heap-profiler implemented.
+    this.profileTypes = [];
+    if (!this.config.disableTime) {
+      this.profileTypes.push(ProfileTypes.Wall);
+      this.timeProfiler =
+          new TimeProfiler(this.config.timeSamplingIntervalMicros);
+    }
   }
 
-  // Begins collection and uploading of profiles.
-  // If profiling fails or another problem is encountered, the returned promise
-  // will be rejected.
-  // TODO: implement
-  // TODO: explain failure situations.
+  /**
+   * Starts and endless loop to poll Stackdriver Profiler server for
+   * instructions, and collects and uploads profiles as requested.
+   * If there is a problem when collecting a profile or uploading a profile to
+   * Stackdriver Profiler, this problem will be logged at the debug level.
+   * If there is a problem polling Stackdriver Profiler for instructions
+   * on the type of profile created, an error will be thrown.
+   */
   async start(): Promise<void> {
-    return Promise.reject(
-        new Error('start() is unimplemented for ProfileAgent'));
+    return this.createAndUploadProfile(0);
+  }
+
+
+  /**
+   * Talks to Stackdriver Profiler server to create profile when instructed
+   * and then upload that profile.
+   *
+   * @param delayMillis - milliseconds to wait before polling Stackdriver
+   * Profiler server to see if it's time to create a profile and profile.
+   */
+  async createAndUploadProfile(delayMillis: number): Promise<void> {
+    this.profilingTimeoutId = setTimeout(async () => {
+      const startCreateMillis = Date.now();
+      const prof = await this.createProfile();
+      await this.profileAndUpload(prof);
+      const endCreateMillis = Date.now();
+      this.createAndUploadProfile(
+          this.config.minTimeBetweenProfilesMillis -
+          (endCreateMillis - startCreateMillis));
+    }, delayMillis);
+    this.profilingTimeoutId.unref();
+  }
+
+  /**
+   * Talks to Stackdriver Profiler server, which hangs until server indicates
+   * job should be profiled.
+   *
+   * If any problem is encountered, an error will be thrown.
+   * TODO: implement backoff and retry when error encountered. createProfile()
+   * should be retried when response indicates this request should be retried
+   * or with exponential backoff (up to one hour) if the response does not
+   * indicate when to retry this request.
+   *
+   * Public to allow for testing.
+   */
+  async createProfile(): Promise<RequestProfile> {
+    const reqBody = {
+      deployment: {
+        projectId: this.config.projectId,
+        target: this.config.serviceContext.service,
+        labels: {zone: this.config.zone, instance: this.config.instance},
+      },
+      profileType: this.profileTypes,
+    };
+    const options = {
+      method: 'POST',
+      uri: '/profiles',
+      body: reqBody,
+      json: true,
+    };
+
+    return this.request(options).then((result) => {
```

Comment:
You're in an async function, so you can `await` rather than `then`ing. Rule of thumb: There should be very few (if any) reasons to use `.then` in this codebase given that we can use `await` now.